Explainer for FragmentDirective API

bokand opened the pull request.

bokand commented:
Also +@tomayac - would value your thoughts if you have some time to look this over.

tomayac commented on fragment-directive-api.md:
```diff
@@ -0,0 +1,211 @@
+
+* Attach comments/responses to specific parts of text on a page or provide helpful UI - e.g. [Marginalia](https://indieweb.org/marginalia)
+
+* Enable pages to easily create text directive links. The rules for how text is matched are [necessarily complicated](https://wicg.github.io/scroll-to-text-fragment/#find-a-range-from-a-text-directive); they must consider word boundaries, DOM node display types and visibility and various nuances of how DOM is traversed. This API allows an author let the browser generate a valid text directive URL ...
```
```suggestion
* Enable pages to easily create text directive links. The rules for how text is matched are [necessarily complicated](https://wicg.github.io/scroll-to-text-fragment/#find-a-range-from-a-text-directive); they must consider word boundaries, DOM node display types and visibility, and various nuances of how DOM is traversed. This API allows an author to let the browser generate a valid text directive URL for a given Range.
```

tomayac commented on fragment-directive-api.md:
```diff
@@ -0,0 +1,211 @@
+
+* Enable pages to easily create text directive links. The rules for how text is matched are [necessarily complicated](https://wicg.github.io/scroll-to-text-fragment/#find-a-range-from-a-text-directive); they must consider word boundaries, DOM node display types and visibility and various nuances of how DOM is traversed. This API allows an author let the browser generate a valid text directive URL ...
+
+* Enable text directives in cross-origin iframes. To prevent [XS-Search attacks](https://wicg.github.io/scroll-to-text-fragment/#example-4d0b486d:~:text=A%20malicious%20page%20embeds%20a%20cross%2Dorigin%20victim%20in%20an%20iframe,its%20own%20document.), text directives are not applied when navigated from a cross-origin initiator. However, an iframe can navigate itself to a text directive. By al ...
```
```suggestion
* Enable text directives in cross-origin iframes. To prevent [XS-Search attacks](https://wicg.github.io/scroll-to-text-fragment/#example-4d0b486d:~:text=A%20malicious%20page%20embeds%20a%20cross%2Dorigin%20victim%20in%20an%20iframe,its%20own%20document.), text directives are not applied when navigated from a cross-origin initiator. However, an iframe can navigate itself to a text directive. By allowing an embedder page to read the text directive, it can `postMessage()` it to a cross-origin document that's opted-in to this behavior, enabling deep linking to the inner frame (see examples section below).
```

tomayac commented on fragment-directive-api.md:
````diff
@@ -0,0 +1,211 @@
+};
+```
+
+Why a SelectorDirective base-class, in addition to Directive? The [proposed](https://github.com/WICG/scroll-to-text-fragment/blob/main/EXTENSIONS.md#proposed-solution) CSS selector directive would behave very similarly to a text directive and allows createSelectorDirective to return a SelectorDirective. OTOH, the proposed [note selector](https://github.com/bokand/web-annotations/blob/main/URL-bas ...
````
```suggestion
Why a `SelectorDirective` base-class, in addition to `Directive`? The [proposed](https://github.com/WICG/scroll-to-text-fragment/blob/main/EXTENSIONS.md#proposed-solution) CSS selector directive would behave very similarly to a text directive and allows `createSelectorDirective()` to return a `SelectorDirective`. OTOH, the proposed [note selector](https://github.com/bokand/web-annotations/blob/main/URL-based-annotation.md) would not fit this interface.
```

tomayac commented on fragment-directive-api.md:
```diff
@@ -0,0 +1,211 @@
+
+Why a SelectorDirective base-class, in addition to Directive? The [proposed](https://github.com/WICG/scroll-to-text-fragment/blob/main/EXTENSIONS.md#proposed-solution) CSS selector directive would behave very similarly to a text directive and allows createSelectorDirective to return a SelectorDirective. OTOH, the proposed [note selector](https://github.com/bokand/web-annotations/blob/main/URL-bas ...
+
+_TODO: Maybe SelectorDirective is unnecessary? Callers could always determine the directive type using `Directive.type` if they need. Also, it may actually make sense for `note` to provide `getMatchingRange`...)_
```
```suggestion
_TODO: Maybe `SelectorDirective` is unnecessary? Callers could always determine the directive type using `Directive.type` if they need to. Also, it may actually make sense for `note` to provide `getMatchingRange()`...)_
```

tomayac commented on fragment-directive-api.md:
````diff
@@ -0,0 +1,211 @@
+```JS
+// Embedder document
+const text_directives =
+    document.fragmentDirectives.items.filter(i => i.type == "text"));
````
```suggestion
    document.fragmentDirectives.items.filter(i => i.type === "text"));
```

tomayac commented on fragment-directive-api.md:
````diff
@@ -0,0 +1,211 @@
+```JS
+//Embedee document
+window.onmessage = (e) => {
+  if (e.type == 'text-directives') {
````
```suggestion
  if (e.type === 'text-directives') {
```

tomayac commented on fragment-directive-api.md:
```diff
@@ -0,0 +1,211 @@
+
+## FragmentDirective.items
+
+The `items` array reflects the currently active directives on the page. Using text directives as an example, an entry should exist in `items` for a text directive as long a highlight is showing. If the user dismisses the highlight it is removed from the array. Conversely, if the directive is removed from `items` programmatically (see next section), the highlight should be removed from the page.
```
```suggestion
The `items` array reflects the currently active directives on the page. Using text directives as an example, an entry should exist in `items` for a text directive as long a highlight is showing. If the user dismisses the highlight, it is removed from the array. Conversely, if the directive is removed from `items` programmatically (see next section), the highlight should be removed from the page.
```

tomayac commented on fragment-directive-api.md:
````diff
@@ -0,0 +1,211 @@
+location.hash = ":~:text=foo%20bar";
+```
+
+Will add a text directive to the page, highlighting "foo bar" and adding a `TextDirective` to `fragmentDirective.items`. However, this still runs the [fragment directive stripping steps](https://wicg.github.io/scroll-to-text-fragment/#process-and-consume-fragment-directive):
````
```suggestion
The snippet above will add a text directive to the page, highlighting "foo bar" and adding a `TextDirective` to `fragmentDirective.items`. However, this still runs the [fragment directive stripping steps](https://wicg.github.io/scroll-to-text-fragment/#process-and-consume-fragment-directive):
```

tomayac commented on fragment-directive-api.md:
```diff
@@ -0,0 +1,211 @@
+There's also the question of what happens to existing directives in `fragmentDirective.items` when the hash is modified. In the cases below, suppose the user navigated to `https://example.org/blog.html#:~:text=acme`.
+
+1. What should happen when script sets a hash with no fragment directive? (e.g. `location.hash = 'page1';`).
+2. What should happen when script sets a hash with an unrelated directive? (e.g. `location.hash = ':~:note(href=notes.example.org);'`)
```
```suggestion
2. What should happen when script sets a hash with an unrelated directive? (e.g. `location.hash = ':~:note(href=notes.example.org)';`)
```

tomayac commented on fragment-directive-api.md:
```diff
@@ -0,0 +1,211 @@
+
+For case 2, it also seems like we shouldn't remove the text directive. Directives of different types should behave independently. That is, adding an annotation to a page shouldn't clear text highlights.
+
+In case 3 either behavior could work: a new highlight should be added and the existing one kept OR the new highlight replaces all existing ones. Though, if additive, it means there's no way to remove directives.
```
```suggestion
In case 3, either behavior could work: a new highlight should be added and the existing one kept OR the new highlight replaces all existing ones. Though, if additive, it means there's no way to remove directives.
```

tomayac reviewed (commented):
I think this goes in the right direction.

- I think the IDL needs to state where it's exposed (`document`).
- There's some redundancy of repeating the directive name `text=` when the type is known (`new TextDirective("text=foo")` vs. a hypothetical simpler `new TextDirective("se")`), but I think that's fine.

bokand commented on fragment-directive-api.md:
```diff
@@ -0,0 +1,211 @@
+
+This document proposes a programmatic API through which authors can interact with text (and future) directives.
+
+Today, when a page is loaded with a text directive such as `https://example.org#:~:text=foo,bar`, the author has no way<sup>1</sup> to tell that a text directive was set or what text was highlighted. The fragment directive portion of the URL (everything in the fragment after and including `:~:`) is stripped from the URL when the document is loaded. This is done for two reasons:
```
Neat! Thanks!

bokand commented:
Thanks Thomas!

> I believe the IDL needs to state where it's exposed (document).

You're right, though it's `[Exposed=Window]` since `Exposed` works on a global object. FragmentDirective is already spec'ed in our draft spec as being a member of `Document` - I added the existing IDL here for clarity. Maybe it'd make sense to also expose `TextDirective` and related on `Worker` as well? `getMatchingRange` couldn't work but the attributes might be useful? Thoughts?

> There's some redundancy of repeating the directive name text= when the type is known (new TextDirective("text=foo") vs. a hypothetical simpler new TextDirective("foo")), but I think that's fine.

Good catch, I have a mix of both styles here - I agree the latter is more ergonomic so I changed the examples to that.

It might be useful to have a `Directive.parse(DOMString)` that just returns a list of `Directive` objects of the right type (e.g. so you could do `Directive.parse("text=foo&note(href=example.com)&selector(...)")` and get back 3 `Directive` objects) but I'm weary of adding too many bells and whistles before we know how it'll be used.

bokand commented:
@flackr please also take a look

flackr commented on fragment-directive-api.md:
````diff
@@ -0,0 +1,230 @@
+
+```JS
+// Coming from a server-side WebMention:
+const comment_text = "Great Point!";
````
How is this use case different from a note in which case the content would be hidden?

Or perhaps more specifically, why would the location of the note be coming from a fragmentdirective but the note content be coming from another server?

bokand replied:
This is describing a separate idea/process to annotations than we're working on - see https://indieweb.org/marginalia. 

In the indieweb world, blog A can make reference to something on blog B and notify blog B by sending it a web mention. If this is in reply to a specific part of Blog B, the webmention needs a way to specify that, i.e. specify what range of text it's replying to. Blog B's server can then go and fetch the content from blog A and attach it to the range specified in the text directive.

marginalia requires some way for blog A to make reference to a specific part of blog B. The indieweb folks envisaged using [fragmention](https://indieweb.org/fragmention) for this but text directives solve the same problem and are actually shipped. In their case, blog A has opted into sharing the reply/annotation with blog B which isn't the case (by default) for our user-generated annotations.

flackr commented on fragment-directive-api.md:
```diff
@@ -0,0 +1,230 @@
+* Enable pages to easily create text directive links. The rules for how text is matched are [necessarily complicated](https://wicg.github.io/scroll-to-text-fragment/#find-a-range-from-a-text-directive); they must consider word boundaries, DOM node display types and visibility, and various nuances of how DOM is traversed. This API allows an author to let the browser generate a valid text directive ...
+
+* Enable text directives in cross-origin iframes. To prevent [XS-Search attacks](https://wicg.github.io/scroll-to-text-fragment/#example-4d0b486d:~:text=A%20malicious%20page%20embeds%20a%20cross%2Dorigin%20victim%20in%20an%20iframe,its%20own%20document.), text directives are not applied when navigated from a cross-origin initiator. However, an iframe can navigate itself to a text directive. By al ...
+
```
Another use case is to provide application specific helpful UI. E.g. a sublime like editor might highlight sections of the preview containing notes, or an application could provide an arrow that points to the fact that there are notes / text to be read further down, possibly also jumping to that note when clicked.

bokand replied:
Great idea - added

flackr commented on fragment-directive-api.md:
```diff
@@ -0,0 +1,230 @@
+};
+
+partial interface Document {
+    [SameObject] readonly attribute FragmentDirective fragmentDirective;
```
Document seems like a strange place to access the current fragmentdirective. Wouldn't these API's to access / modify the fragment directive make more sense as part of Location?

bokand replied:
Heh - indeed. Alas, this didn't work out. See crbug.com/1057795 and https://github.com/whatwg/html/issues/5523 for some context.

`window.location` has a bunch of legacy baggage and behaves in some surprising ways that can lead to leaking objects like this across origins if we're not careful; `Location` needs to refer to the active document in the browsing context rather than its own.

bokand commented:
@flackr Since I'll be out for the next 6 weeks I'm going to merge the latest changes so I have something to point to in the intent to prototype but feel free to make changes while I'm gone if you have any.
